### PR TITLE
Enable gamestore to offers services/items for tournament coins

### PR DESCRIPTION
**Pull request by bosmak**
# Description

Allows offers tagged with `coinType = GameStore.CointType.Tournament` to correctly charge Tournament Coins.
Added correct history for the store transactions following the coin type.
Removed duplicate charge and history when changing the Name or Sex of a Hireling.

### **Actual**

Buying something tagged with Tournament Coins will be charged on Tibia Coins.

Buying `Hireling Sex Change` or `Hireling Name Change` charges twice the amount and add two purchases to the history.

### **Expected**

Buying something tagged with Tournament Coins will be charged using Tournament Coins.

Buying `Hireling Sex Change` or `Hireling Name Change` charges the displayed amount and adds a single purchase to the history.

## Fixes

\# (issue)

## Type of change

Please delete options that are not relevant.

  - [X] Bug fix (non-breaking change which fixes an issue)
  - [X] New feature (non-breaking change which adds functionality)

## How Has This Been Tested

On `data/modules/scripts/gamestore/gamestore.lua` change the line 4492 from `offer.coinType = GameStore.CointType.Coin` to `offer.coinType = GameStore.CointType.Tournament` to change all offers to use Tournament Coins.

Go inside the game and buy one of each OfferType; check history.

**Test Configuration**:

  - Server Version: Canary@5a6da6c1436be0d4a0a9b06a7a9d2fb98325eb3f
  - Client: 12.86.11871
  - Operating System: Ubuntu:21.04

## Checklist

  - [X] My code follows the style guidelines of this project
  - [X] I have performed a self-review of my own code
  - [ ] I checked the PR checks reports
  - [ ] I have commented my code, particularly in hard-to-understand areas
  - [ ] I have made corresponding changes to the documentation
  - [ ] My changes generate no new warnings
  - [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by QuebradaZN** [comment]
After making a purchase using Tournament Coins, you are not debited from Tournament Points but from Tibia Coins

**Review by majestyotbr — APPROVED** [review]

**Review by majestyotbr — APPROVED** [review]

File: data/migrations/19.lua [hunk 1]
```diff
@@ -1,3 +1,9 @@
 function onUpdateDatabase()
-    return false -- true = There are others migrations file | false = this is the last migration file
+	Spdlog.info("Updating database to version 20 (Gamestore accepting Tournament Coins)")
+
+	db.query("ALTER TABLE `accounts` ADD `tournament_coins` int(11) NOT NULL DEFAULT 0 AFTER `coins`")
+	db.query("ALTER TABLE `store_history` ADD `coin_type` tinyint(1) NOT NULL DEFAULT 0 AFTER `description`")
+    db.query("ALTER TABLE `store_history` DROP COLUMN `coins`") -- Not in use anywhere.
```
Comment (dudantas) [inline data/migrations/19.lua]:
```suggestion
	db.query("ALTER TABLE `store_history` DROP COLUMN `coins`") -- Not in use anywhere.
```

File: data/migrations/20.lua [hunk 2]
```diff
@@ -0,0 +1,3 @@
+function onUpdateDatabase()
+    return false -- true = There are others migrations file | false = this is the last migration file
+end
```
Comment (dudantas) [inline data/migrations/20.lua]:
```suggestion
end

```

File: data/modules/scripts/gamestore/init.lua [hunk 3]
```diff
@@ -366,8 +367,10 @@ function parseBuyStoreOffer(playerId, msg)
 
 	-- At this point the purchase is assumed to be formatted correctly
 	local offerPrice = offer.type == GameStore.OfferTypes.OFFER_TYPE_EXPBOOST and GameStore.ExpBoostValues[player:getStorageValue(GameStore.Storages.expBoostCount)] or offer.price
-
-	if not player:canRemoveCoins(offerPrice) then
+	local offerCoinType = offer.coinType
+
```
Comment (dudantas) [inline data/modules/scripts/gamestore/init.lua]:
```suggestion

```

File: data/modules/scripts/gamestore/init.lua [hunk 4]
```diff
@@ -907,10 +911,10 @@ function sendUpdateCoinBalance(playerId)
 	msg:addByte(GameStore.SendingPackets.S_CoinBalance)
 	msg:addByte(0x01)
 
-	msg:addU32(player:getCoinsBalance())
-	msg:addU32(player:getCoinsBalance())
-	msg:addU32(player:getCoinsBalance())
-	msg:addU32(0) -- Tournament Coins
+	msg:addU32(player:getCoinsBalance())      -- Tibia Coins
+	msg:addU32(player:getCoinsBalance()) 	  -- How many are Transferable
+	msg:addU32(0) 							  -- How many are reserved for a Character Auction
```
Comment (dudantas) [inline data/modules/scripts/gamestore/init.lua]:
```suggestion
	msg:addU32(player:getCoinsBalance()) -- Tibia Coins
	msg:addU32(player:getCoinsBalance()) -- How many are Transferable
	msg:addU32(0) -- How many are reserved for a Character Auction
```

File: data/modules/scripts/gamestore/init.lua [hunk 5]
```diff
@@ -1512,8 +1516,8 @@ function GameStore.processHirelingPurchase(player, offer, productType, hirelingN
 			return error({code = 1, message = "Error delivering your hireling lamp, try again later."})
 		end
 
-		player:removeCoinsBalance(offer.price)
-		GameStore.insertHistory(player:getAccountId(), GameStore.HistoryTypes.HISTORY_TYPE_NONE, offer.name .. ' ('.. hirelingName ..')', (offer.price) * -1)
+		player:makeCoinTransaction(offer, hirelingName)
+
```
Comment (dudantas) [inline data/modules/scripts/gamestore/init.lua]:
```suggestion

```

File: data/modules/scripts/gamestore/init.lua [hunk 6]
```diff
@@ -1604,6 +1610,74 @@ function Player.addCoinsBalance(self, coins, update)
 	return true
 end
 
+--- Tournament Coins
+function Player.getTournamentBalance(self)
+	resultId = db.storeQuery("SELECT `tournament_coins` FROM `accounts` WHERE `id` = " .. self:getAccountId())
+	if not resultId then return 0 end
```
Comment (dudantas) [inline data/modules/scripts/gamestore/init.lua]:
```suggestion
	if not resultId then
		return 0
	end
```

File: data/modules/scripts/gamestore/init.lua [hunk 7]
```diff
@@ -1761,18 +1835,20 @@ function HandleHirelingNameChange(playerId, offer, newHirelingName)
...
+		end
+
 		local lamp = player:findHirelingLamp(hireling:getId())
 		if lamp then
 			lamp:setAttribute(ITEM_ATTRIBUTE_DESCRIPTION, "This mysterious lamp summons your very own personal hireling.\nThis item cannot be traded.\nThis magic lamp is the home of " .. hireling:getName() .. ".")
 		end
-		GameStore.insertHistory(player:getAccountId(), GameStore.HistoryTypes.HISTORY_TYPE_NONE, offer.name .. ' ('.. oldName .. '->' .. newHirelingName ..')', (offer.price) * -1)
-
-		player:showInfoModal('Info',string.format('%s has been renamed to %s', oldName, newHirelingName))
+
```
Comment (dudantas) [inline data/modules/scripts/gamestore/init.lua]:
```suggestion

```

File: data/modules/scripts/gamestore/init.lua [hunk 8]
```diff
@@ -1761,18 +1835,20 @@ function HandleHirelingNameChange(playerId, offer, newHirelingName)
...
 		local lamp = player:findHirelingLamp(hireling:getId())
 		if lamp then
 			lamp:setAttribute(ITEM_ATTRIBUTE_DESCRIPTION, "This mysterious lamp summons your very own personal hireling.\nThis item cannot be traded.\nThis magic lamp is the home of " .. hireling:getName() .. ".")
 		end
-		GameStore.insertHistory(player:getAccountId(), GameStore.HistoryTypes.HISTORY_TYPE_NONE, offer.name .. ' ('.. oldName .. '->' .. newHirelingName ..')', (offer.price) * -1)
-
-		player:showInfoModal('Info',string.format('%s has been renamed to %s', oldName, newHirelingName))
+		
+		--player:showInfoModal('Info',string.format('%s has been renamed to %s', oldName, newHirelingName))
```
Comment (dudantas) [inline data/modules/scripts/gamestore/init.lua]:
```suggestion
		Spdlog.debug(string.format('%s has been renamed to %s', oldName, newHirelingName))
```

File: data/modules/scripts/gamestore/init.lua [hunk 9]
```diff
@@ -1809,9 +1885,8 @@ function HandleHirelingSexChange(playerId, offer)
 		hireling.sex = changeTo
 		hireling.looktype = lookType
 
-		GameStore.insertHistory(player:getAccountId(), GameStore.HistoryTypes.HISTORY_TYPE_NONE, offer.name .. ' ('.. hireling:getName() ..')', (offer.price) * -1)
-
-		player:showInfoModal('Info',string.format('%s sex was changed to %s', hireling:getName(), sexString))
+		--player:showInfoModal('Info',string.format('%s sex was changed to %s', hireling:getName(), sexString))
```
Comment (dudantas) [inline data/modules/scripts/gamestore/init.lua]:
```suggestion
		Spdlog.debug(string.format('%s sex was changed to %s', hireling:getName(), sexString))
```

File: schema.sql [hunk 10]
```diff
@@ -23,16 +23,17 @@ INSERT INTO `server_config` (`config`, `value`) VALUES ('db_version', '19'), ('m
...
+  `name`                varchar(32)   NOT NULL,
+  `password`            char(40)      NOT NULL,
+  `email`               varchar(255)  NOT NULL DEFAULT '',
+  `premdays`            int(11)       NOT NULL DEFAULT '0',
+  `lastday`             int(10)       UNSIGNED NOT NULL DEFAULT '0',
+  `type`                tinyint(1)    UNSIGNED NOT NULL DEFAULT '1',
+  `coins`               int(12)       UNSIGNED NOT NULL DEFAULT '0',
+  `tournament_coins`    int(12)       UNSIGNED NOT NULL DEFAULT '0',
+  `creation`            int(11)       UNSIGNED NOT NULL DEFAULT '0',
+  `recruiter`           INT(6)        DEFAULT 0,
```
Comment (dudantas) [inline schema.sql]:
```suggestion
  `id` int(11) UNSIGNED NOT NULL AUTO_INCREMENT,
  `name` varchar(32) NOT NULL,
  `password` char(40) NOT NULL,
  `email` varchar(255) NOT NULL DEFAULT '',
  `premdays` int(11) NOT NULL DEFAULT '0',
  `lastday` int(10) UNSIGNED NOT NULL DEFAULT '0',
  `type` tinyint(1) UNSIGNED NOT NULL DEFAULT '1',
  `coins` int(12) UNSIGNED NOT NULL DEFAULT '0',
  `tournament_coins` int(12) UNSIGNED NOT NULL DEFAULT '0',
  `creation` int(11) UNSIGNED NOT NULL DEFAULT '0',
  `recruiter` INT(6) DEFAULT 0,
```

File: schema.sql [hunk 11]
```diff
@@ -865,20 +866,21 @@ CREATE TABLE IF NOT EXISTS `player_storage` (
...
+	`coin_amount` int(12) NOT NULL,
+	`time`        bigint(20) UNSIGNED NOT NULL,
+	`timestamp`   int(11) NOT NULL DEFAULT '0',
+	INDEX `account_id` (`account_id`),
+	CONSTRAINT `store_history_pk`
+		PRIMARY KEY (`id`),
+	CONSTRAINT `store_history_account_fk`
+		FOREIGN KEY (`account_id`) REFERENCES `accounts` (`id`)
+		ON DELETE CASCADE
+) ENGINE = InnoDB DEFAULT CHARSET = utf8;
```
Comment (dudantas) [inline schema.sql]:
```suggestion
	`id` int(11) NOT NULL AUTO_INCREMENT,
	`account_id` int(11) UNSIGNED NOT NULL,
	`mode` smallint(2) NOT NULL DEFAULT '0',
	`description` varchar(3500) NOT NULL,
	`coin_type` tinyint(1) NOT NULL DEFAULT '0',
	`coin_amount` int(12) NOT NULL,
	`time` bigint(20) UNSIGNED NOT NULL,
	`timestamp` int(11) NOT NULL DEFAULT '0',
	INDEX `account_id` (`account_id`),
	CONSTRAINT `store_history_pk`
		PRIMARY KEY (`id`),
	CONSTRAINT `store_history_account_fk`
		FOREIGN KEY (`account_id`) REFERENCES `accounts` (`id`)
		ON DELETE CASCADE
) ENGINE = InnoDB DEFAULT CHARSET = utf8;
```

**Review by dudantas — CHANGES_REQUESTED** [review]
Some suggestions related to indentation. The code itself seems to be ok.

**Review by dudantas — APPROVED** [review]

**Review by majestyotbr — APPROVED** [review]

**Review by guispiller — APPROVED** [review]

**Review by JoaoCRDias — APPROVED** [review]

**Review by QuebradaZN — APPROVED** [review]
Testing and apparently everything is ok!

**Pull request merged**